File: doc/tutorial/code/dc1d/dc1dsmooth.cpp
/***************************************************************************
 * This file belongs to GIMLi (Geophysical Inversion and Modelling) library*
 * and was created for documentation purposes only.                        *
 * DC1dSMOOTH is a smooth inversion of resistivity soundings               *
 * Run with example file using: dc1dsmooth sond1-100dat                    *
***************************************************************************/

#include <optionmap.h>
#include <inversion.h>
#include <dc1dmodelling.h>
#include <string>

using namespace GIMLI;

int main( int argc, char *argv [] )
{
    std::string dataFile( NOT_DEFINED );
    double errPerc = 3.0, lambda = 20.0;
    int nlay = 30;
    bool verbose = true, lambdaOpt = false, isBlocky = false, optimizeChi1 = false;
    OptionMap oMap( "DC1dsmooth - smooth 1d dc resistivity inversion" );
    oMap.add( lambdaOpt,    "O" , "OptimizeLambda", "Optimize model smoothness using L-curve." );
    oMap.add( isBlocky,     "B" , "BlockyModel"   , "Blocky (L1) model constraints." );
    oMap.add( optimizeChi1, "C" , "OptimizeChi1"  , "Optimize lambda subject to chi^2=1." );
    oMap.add( lambda,       "l:", "lambda"        , "Regularization strength lambda[100]." );
    oMap.addLastArg( dataFile, "Datafile" );
    oMap.parse( argc, argv );

    /*! Data: read data file from column file */
    RMatrix abmnr; 
    loadMatrixCol( abmnr, dataFile ); //! read data
    RVector ab2(  abmnr[ 0 ] );        //! first column
    RVector mn2(  abmnr[ 1 ] );        //! second column
    RVector rhoa( abmnr[ 2 ] );        //! third column

    /*! Define discretization according to AB/2 */
    double maxDep = max( ab2 ) / 2;   //! rule of thumb
    std::cout << "Maximum depth estimated to " << maxDep << std::endl;
    RVector thk( nlay - 1, maxDep / ( nlay - 1 ) );
    thk *= ( maxDep / sum( thk ) / 3 );

    /*! Transformations: log for app. resisivity and thickness, logLU for resistivity */
//    RTransLogLU transRho( lbound, ubound ); // alternative
    RTransLog transRho;
    RTransLog transRhoa;

    /*! Forward operator, transformations and constraints */
    DC1dRhoModelling f( thk, ab2, mn2 );
//    f.region( 0 )->setTransModel( transRho );

    /*! Set up inversion with full matrix, data and forward operator */
    RInversion inv( rhoa, f, verbose );
    inv.setTransData( transRhoa );
    inv.setTransModel( transRho );
    inv.setRelativeError( errPerc / 100.0 );
    inv.setLambda( lambda );
    inv.setOptimizeLambda( lambdaOpt );
    inv.setBlockyModel( isBlocky );
    inv.stopAtChi1( false );
    RVector model( nlay, median( rhoa ) );
    inv.setModel( model );

    if ( optimizeChi1 ) {
        model = inv.runChi1( 0.1 );
    } else {
        model = inv.run();
    }
    if ( verbose ) std::cout << "model = " << model << std::endl;
    save( model, "resistivity.vec" );
    save( thk, "thickness.vec" );

    return EXIT_SUCCESS;
}

File: doc/tutorial/code/dc1d/dc1dsmooth.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pygimli as g
import pylab as P
from pygimli.utils.base import draw1dmodel

nlay = 30
lam = 20.
errPerc = 3.
isBlocky = False

abmnr=g.RMatrix()
g.loadMatrixCol(abmnr,"sond1-100.ves")
ab2  = abmnr[0]
mn2  = abmnr[1]
rhoa = abmnr[2]

maxDep = max( ab2 ) / 2.
print "Maximum depth estimated to ", maxDep
thk = g.RVector( nlay - 1, maxDep / ( nlay - 1 ) )
thk *= ( maxDep / sum( thk ) / 3 )

transRho  = g.RTransLog()
transRhoa = g.RTransLog()

f=g.DC1dRhoModelling( thk, ab2, mn2 )

inv = g.RInversion( rhoa, f, True )
model = g.RVector( nlay, P.median( rhoa ) )
inv.setModel( model )
inv.setTransData( transRhoa )
inv.setTransModel( transRho )
inv.setRelativeError( errPerc / 100.0 )
inv.setLambda( lam )
model = inv.run()

model2 = g.RVector( nlay, P.median( rhoa ) )
inv.setModel( model2 )
inv.setBlockyModel( True )
model2 = inv.run()

fig = P.figure(1)
fig.clf()
ax1=fig.add_subplot(121)
P.loglog( rhoa, ab2, 'rx-', inv.response(), ab2, 'b-' )
P.axis('tight')
P.ylim((max(ab2),min(ab2)))
P.grid(which='both')
P.xlabel(r"\rho_a in \Omegam");
P.ylabel("AB/2 in m");
P.legend(("measured", "fitted"),loc="upper left");
ax2=fig.add_subplot(122)
draw1dmodel(model,thk,r'\rho in \Omega m')
draw1dmodel(model2,thk,r'\rho in \Omega m')
draw1dmodel([100, 500, 20, 1000],[0.5,3.5,6])
P.show()